### DISCUSSION TITLE
Unable to disable rich output without uninstalling rich

### DISCUSSION THREAD

**Discussion by graygreenblue** [fastapi/typer#578] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer

app = typer.Typer()

@app.command()
def hello(name: str):
    print(f"Hello {name}!")

app()
```

### Description

I'm trying to *have the option* to disable the rich output without having to uninstall rich or use a venv without rich.

I've tried the "rich_markup_mode=None" and "rich_help_panel=None"

None of these seem to disable rich output.

### Operating System

Windows

### Typer Version

0.7.0

### Python Version

3.10.10

**Comment by nandgator** [comment]
With some help, I've created a patch script:

<details><summary><b><code>patch.sh</code></b></summary>

```bash
#!/bin/bash

print_help() {
    # Print the help text
    echo "
    Usage: $0 <filename> <content> <line_number>
           $0 <filename> <operation> <start_line> <end_line>

    Arguments
      <filename>      : File name
      <operation>     : Operation to perform
      <start_line>    : Start line number
      <end_line>      : End line number
      <content>       : Content to be injected
      <line_number>   : Line number to be injected into

    Description
      This script performs operations on files based on the provided arguments.
      In case of 3 arguments, <filename> is the 'file name', <content> is a 'string',
      and <line_number> is a 'line number'.
      In case of 4 arguments, <filename> is the 'file name', <operation> is the
      'operation' to perform, <start_line> is the 'starting line number', and
      <end_line> is the 'ending line number'.
    "
}

if [ "$#" -gt 2 ]; then
    file_path="$1"
    # Check if the file exists
    if [ ! -f "${file_path}" ]; then
        echo "File '${file_path}' does not exist."
        exit 1
    fi
fi

# Check the number of arguments
if [ "$#" -eq 3 ]; then
    # Process the case with 3 arguments
    content="$2"
    line_number="$3"

    inject() {
        # sed inserts in location: line_number + 1, therefore
        sed -i "$((${line_number} - 1)) a ${content}" ${file_path} &&
            echo "Injected code into ${file_path}:${line_number}."
    }

    inject

elif [ "$#" -eq 4 ]; then
    # Process the case with 4 arguments
    operation="$2"
    start_line="$3"
    end_line="$4"

    comment() {
        # Comment specific lines
        sed -i "${start_line},${end_line}s/^/#/" "${file_path}" &&
            echo "Lines $start_line to $end_line commented in ${file_path}."
    }

    uncomment() {
        # Uncomment specific lines
        sed -i "${start_line},${end_line}s/^# *//" "${file_path}" &&
            echo "Lines $start_line to $end_line uncommented in ${file_path}."
    }

    if [ "${operation}" == "c" ]; then
        comment
    elif [ "${operation}" == "u" ]; then
        uncomment
    fi

else
    print_help
    if [ "$#" -ne 0 ]; then
        echo "E: Bad options provided."
        exit 1
    fi
    exit 0

fi
```

</details>

You can comment and inject code, resulting in getting `rich` disabled.

```console
$ ./patch.sh .venv/lib/python3.11/site-packages/typer/core.py c 35 41
Lines 35 to 41 commented in .venv/lib/python3.11/site-packages/typer/core.py.
$  ./patch.sh .venv/lib/python3.11/site-packages/typer/core.py 'rich = None # type: ignore' 42
Injected code into .venv/lib/python3.11/site-packages/typer/core.py:42.
$ ./patch.sh .venv/lib/python3.11/site-packages/typer/main.py c 39 47
Lines 39 to 47 commented in .venv/lib/python3.11/site-packages/typer/main.py.
$  ./patch.sh .venv/lib/python3.11/site-packages/typer/main.py 'rich = None # type: ignore' 48
Injected code into .venv/lib/python3.11/site-packages/typer/main.py:48.
```

This of course depends upon getting path and the line numbers correct. So keep an eye on [`core.py`](https://github.com/tiangolo/typer/blob/master/typer/core.py#L35-L41) & [`main.py`](https://github.com/tiangolo/typer/blob/master/typer/core.py#L39-L47)

<details><summary>Before</summary>

https://github.com/tiangolo/typer/blob/f83c66d539b95d00de74e7de6a14b5c53ae0bfe6/typer/core.py#L35-L41

https://github.com/tiangolo/typer/blob/f83c66d539b95d00de74e7de6a14b5c53ae0bfe6/typer/main.py#L39-L47

</details>

<details><summary>After</summary>

`/../core.py`

```py
#try:
#    import rich
#
#    from . import rich_utils
#
#except ImportError:  # pragma: nocover
#    rich = None  # type: ignore
rich = None # type: ignore
```

`/../main.py`

```py
#try:
#    import rich
#    from rich.console import Console
#    from rich.traceback import Traceback
#
#    console_stderr = Console(stderr=True)
#
#except ImportError:  # pragma: nocover
#    rich = None  # type: ignore
rich = None # type: ignore
```

**Comment by corneliusroemer** [comment]
What do you mean by "disabling rich output"? Is it the rich tracebacks that bother you? Or rich output in general? Tracebacks can be switched off, see https://stackoverflow.com/questions/76375307/how-to-make-typer-traceback-look-normal

**Reply by nandgator** [reply]
OP means, if `rich` is installed, it will be picked up by default, giving no room for the developer, to switch off 'rich' output (help panels', etc.) in general.

**Reply by 4l1fe** [reply]
I'd also like to disable the rich's feature - **Panel**. 
I don't like a frame around the `--help` output sections  but still need to have it colorized.
![image](https://github.com/tiangolo/typer/assets/21095447/d4b2bf88-93d2-443e-899c-311a309ec9ac)


Is there a way to say to Rich not to use **Panel**?

**Reply by nandgator** [reply]
@4l1fe, you do have these options for the `Typer` class:

https://github.com/tiangolo/typer/blob/f83c66d539b95d00de74e7de6a14b5c53ae0bfe6/typer/main.py#L136-L141

But if the `rich` library is installed in your virtual environment same as `Typer`, it'll be picked up by default.

**Reply by 4l1fe** [reply]
> @4l1fe, you do have these options for the `Typer` class:

And what is a value of `rich_help_panel` to hide panels? Those options are undocumented...

**Reply by 4l1fe** [reply]
Sry, found info about the option here [Help Panels](https://typer.tiangolo.com/tutorial/commands/help/#help-panels)

Sadly, it's not about disabling panels at all.

**Reply by nandgator** [reply]
That's right. A workaround is you can follow the logic of my script as in https://github.com/tiangolo/typer/discussions/578#discussioncomment-6025909

**Reply by 4l1fe** [reply]
I mocked the Panel box to remove the round border as following

```python
import typer


app = typer.Typer()


@app.command()
def create(username: str):
    print(f"Creating user: {username}")


@app.command()
def delete(username: str):
    print(f"Deleting user: {username}")


if __name__ == "__main__":
    from unittest.mock import patch
    from rich.panel import Panel as _Panel
    from rich.box import SIMPLE

    class Panel(_Panel):
        """Replacer of the hardcoded Panel Box type"""

        def __init__(self, *args, **kwargs):
            super().__init__(*args, box=SIMPLE, **kwargs)

    with patch('typer.rich_utils.Panel', Panel):
        app()

```
![image](https://github.com/tiangolo/typer/assets/21095447/e962b074-d05a-46a3-a4ab-377415c477fc)


I do it sometimes. If you are interested in it you can read [the bigger explanation](https://www.kvdm.dev/articles/simply-patch-what-is-unextendable/) about it with an example. Give it a star if you find it helpful 🙂

**Reply by corneliusroemer** [reply]
That's great @4l1fe :) It could be useful to allow configuration of typer's Rich use through config, or wrap your solution a little more magically/abstracted, so one can simply import, as `typer-rich-config` and then do `app = typer_rich_config.Typer(rich_config=config_kwarg)`

**Reply by 4l1fe** [reply]
If you go that far then why not =)

**Comment by coreegor** [comment]
Hello @tiangolo 
There are several issues with using Rich in Typer.
See 
1. current discussion
2. https://github.com/tiangolo/typer/discussions/622
3. https://stackoverflow.com/questions/76601266/disable-rich-help-panel-in-typer

As for me, I'm trying to use Rich library in my project with Typer, but I don't prefer default help formatting through rich. So I think it's only fair to allow the user to set up rich integrations. On the other hand, I also understand that creating an interface that allows the user to modify a rich formatting is not a quick and easy thing to do. So, the fastest way to solve the problem is to add option for disabling formatting through Rich.

Right now I'm faced with the choice of either abandoning Rich or switching to Click. But I don't want to do it because I love Typer.

Can you give this problem some time?

**Reply by JacobKochems** [reply]
>Right now I'm faced with the choice of either abandoning Rich or switching to Click.

Hello @coreegor, lets say you decide to abandon Rich and want to distribute a package. Does this mean whoever has Rich and your package installed would experience the very issues you're trying to avoid?

**Reply by coreegor** [reply]
Hello @JacobKochems 

Yes, it does. But my package is **not intended** for further use in other packages. It is a ready-to-use tool in a specific environment.

**Reply by JacobKochems** [reply]
Hi, I made a pull request (https://github.com/tiangolo/typer/pull/647) with an implementation of a feature similar to the one described by @anthonyraf in #622. Maybe it helps your case and you could review it.

**Reply by coreegor** [reply]
So #647 works for me and this seems to be the solution for this discussion. Thank you!

**Reply by JacobKochems** [reply]
Could you write a review of my PR in #647 according to [this](https://typer.tiangolo.com/help-typer/#review-pull-requests) ?

I just realized that I also turned off the Rich formatting for tracebacks. I added a separate switch for that.

**Comment by tiangolo** [comment, marked as answer]
Thanks for the feedback everyone! This was implemented here: https://github.com/fastapi/typer/pull/859, available since Typer 0.12.5. 

Would that solve your problem? :nerd_face:
